### ISSUE TITLE
hipIpcGetMemHandle fails for memory of arbitrary sizes

### ISSUE THREAD

**Issue by ex-rzr** [ROCm-Developer-Tools/HIP#2257]
`hipIpcGetMemHandle` fails with `hipErrorInvalidDevicePointer` for memory of some sizes but works ok for other sizes:
```
/opt/rocm/bin/hipcc -o ipc_get_mem_handle ipc_get_mem_handle.cpp && ./ipc_get_mem_handle

hipIpcGetMemHandle:          1 0x7f525a000000 = 0
hipIpcGetMemHandle:         65 0x7f5259c00000 = 0
hipIpcGetMemHandle:        123 0x7f5259800000 = 0
hipIpcGetMemHandle:    1048576 0x7f5259400000 = 0
hipIpcGetMemHandle:    2097152 0x7f5259000000 = 0
hipIpcGetMemHandle:   16777216 0x7f4e4b800000 = 0
hipIpcGetMemHandle:    2267136 0x7f5258a00000 = 17
hipIpcGetMemHandle:    2097275 0x7f5258400000 = 17
hipIpcGetMemHandle:    2096807 0x7f4e4b400000 = 0
hipIpcGetMemHandle:   16764871 0x7f4e4a200000 = 17
hipIpcGetMemHandle:   16789561 0x7f4e48e00000 = 17
```

`cudaIpcGetMemHandle` does not care about sizes:
```
nvcc -o ipc_get_mem_handle ipc_get_mem_handle.cu && ./ipc_get_mem_handle

cudaIpcGetMemHandle:          1 0x7fb24f000000 = 0
cudaIpcGetMemHandle:         65 0x7fb24f000200 = 0
cudaIpcGetMemHandle:        123 0x7fb24f000400 = 0
cudaIpcGetMemHandle:    1048576 0x7fb24f000600 = 0
cudaIpcGetMemHandle:    2097152 0x7fb24f200000 = 0
cudaIpcGetMemHandle:   16777216 0x7fb250e00000 = 0
cudaIpcGetMemHandle:    2267136 0x7fb24f400000 = 0
cudaIpcGetMemHandle:    2097275 0x7fb24f800000 = 0
cudaIpcGetMemHandle:    2096807 0x7fb24fc00000 = 0
cudaIpcGetMemHandle:   16764871 0x7fb24a000000 = 0
cudaIpcGetMemHandle:   16789561 0x7fb248000000 = 0
```

[ipc_get_mem_handle.zip](https://github.com/ROCm-Developer-Tools/HIP/files/6270106/ipc_get_mem_handle.zip)

I've found out experimentally that it works correctly when the size is aligned by 2 MB.

There is a note at https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#interprocess-communication:
> Note that allocations made by cudaMalloc() may be sub-allocated from a larger block of memory for performance reasons. In such case, CUDA IPC APIs will share the entire underlying memory block which may cause other sub-allocations to be shared, which can potentially lead to information disclosure between processes. To prevent this behavior, it is recommended to only share allocations with a 2MiB aligned size. 

So I assume that 2 MB is shared memory granularity, i.e a restriction of OS.
https://github.com/ROCm-Developer-Tools/ROCclr/blob/main/device/rocm/rocdevice.cpp#L1964 aligns the size up, but I think, by 4 KB.

Is it safe to always align allocation size by 2 MB? If not, how can I calculate the safe alignment?

If this is a correct behavior and won't be fixed, it should be mentioned somewhere in the documentation.

**Comment by tcgu-amd** [comment]
Hi @ex-rzr, thanks for reaching out! Can you provide your system configuration please? Are you using CUDA or ROCm? (The link https://github.com/ROCm-Developer-Tools/ROCclr/blob/main/device/rocm/rocdevice.cpp#L1964 points to source code for ROCm , or AMD devices, while https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#interprocess-communication points to CUDA). Thanks!

**Comment by ex-rzr** [comment]
ROCm (AMD GPU). The link to CUDA docs is my attempt to find an explanation why it may happen.

**Comment by tcgu-amd** [comment]
Thanks for the additional info @ex-rzr. I ran your reproducer but wasn't able to reproduce your issue. Here is the output:
```
hipIpcGetMemHandle:          1 0x703ad6400000 = 0
hipIpcGetMemHandle:         65 0x703ad6000000 = 0
hipIpcGetMemHandle:        123 0x703ad5c00000 = 0
hipIpcGetMemHandle:    1048576 0x703ad5800000 = 0
hipIpcGetMemHandle:    2097152 0x703ad5400000 = 0
hipIpcGetMemHandle:   16777216 0x703acee00000 = 0
hipIpcGetMemHandle:    2267136 0x703ad4e00000 = 0
hipIpcGetMemHandle:    2097275 0x703ad4800000 = 0
hipIpcGetMemHandle:    2096807 0x703ad4400000 = 0
hipIpcGetMemHandle:   16764871 0x703acdc00000 = 0
hipIpcGetMemHandle:   16789561 0x703acc800000 = 0
```
I am running the latest ROCm 6.2.2 on Ubuntu 24.04 with a Radeon 7900 XT graphics card. 

Would you mind letting us know your combination of OS, ROCm version, and GPU so we can give a try? Thanks!

**Comment by ex-rzr** [comment]
> Would you mind letting us know your combination of OS, ROCm version, and GPU

Honestly, I don't remember, the issue is 3 years old. The GPU was MI100 or/and MI210.

> I ran your reproducer but wasn't able to reproduce your issue.

Good! So I assume it's fixed in the recent versions of ROCm.

**Comment by tcgu-amd** [comment]
> Honestly, I don't remember, the issue is 3 years old. The GPU was MI100 or/and MI210.

Yeah, sorry again for not replying for this long...and thank you for the prompt response and willingness to provide additional information for us even after this many years! We are working on improving community support and we have been revisiting a lot of old open issues recently. Hopefully things will improve from now on.

>Good! So I assume it's fixed in the recent versions of ROCm.

That is likely the case I think. ROCm has been developing really rapidly and a lot of things get introduced and fixed in short periods. I will try to see if I can find how it is fixed. In meantime, I hope you don't mind if I close the ticket for now. Thanks!

**Comment by tcgu-amd** [comment]
@ex-rzr Hey turns out this is a fairly recent fix part of [this commit](https://github.com/ROCm/clr/commit/fa1b7925cf8daf296831993e79f6bd86231ce169). It shipped with ROCm 5.6.0x has has been there since then. Hope this helps!

**Comment by ex-rzr** [comment]
Great, thanks!
